torch/csrc/autograd/autograd_meta.cpp

```diff
@@ -85,6 +85,9 @@ namespace {
         return false;
       }
     }
+    if (base._storage_numel() != other._storage_numel()) {
```

Comment:
By the way, there's a huge performance problem with `_new_zeros_with_same_feature_meta` that arises with the invariant that `base._storage_numel() == other._storage_numel()`

Let's say we have `x: Tensor[5], y: Tensor[3]` and we construct a dual tensor `make_dual(primal=x[0], tangent=BatchedTensor(y))`.

When we set `BatchedTensor(y)` to be the tangent, we require it to have the same storage numel as primal. So we end up creating coercing the tangent into a tensor of shape [3, 5].

If the user never actually uses as_strided on their dual tensor or does in-place operations, then they are paying for memory that they didn't ask for.

How important is it that we add this check? I want to say that forward-mode AD does not need this check during composition with vmap.

Reply:
Thanks for the example, it does seem like it would be an easy way for users to shoot themselves in the foot.

Looks like (A) is already unsupported by vmap as well (in core and in functorch)? In that case I would agree that forward AD does not need this check in composition with vmap.

```
from functorch import vmap
from torch._vmap_internals import _vmap

x = torch.rand(10, 2).select(1, 0)

# Both fail
vmap(lambda x: x.as_strided([1], [1], 1))(x)
_vmap(lambda x: x.as_strided([1], [1], 1))(x)
```

Also filed this issue regarding the error message: https://github.com/pytorch/functorch/issues/308

Reply:
> Looks like (A) is already unsupported by vmap as well (in core and in functorch)?

Nice catch! I forgot we had those checks